# Zero To Hero Big Data Prepration
Taking Advantage of Cloud Technologies to Create Big Data Solutions
    
    AUTHOR: Dr. Roy Jafari 

# Chapter 3: Using the right data strucutres 

## Challenge 2 - lits or linked list?

Linked lists do not have a lot of big data applications, however, they can have magical effects on the situations in which we want to apply in-place changes to a collection of items. In this challenge, we will consider the run time of in-place changes to a typical python list and to its identical linked list. Answer the following questions and perform the following steps.

1.	The following code uses `np.random.randint()` function to create a list with 100,000 random numbers between 0, and 99. Run the code to create a_list. 

In [1]:
import numpy as np
a_list = np.random.randint(0,100,100000).tolist()

2.	Now we want to create the identical linked list version of `a_list`. First, we need to create the class of objects NodeList. The following code defines this class.

In [3]:
class ListNode(object):
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next

Now, we can use this class to create the linked list version of `a_list`. Run the following code to do this.


In [4]:
node = ListNode()
dummy = node
for n in a_list:
    node.next = ListNode(n)
    node = node.next  
head_linked_list = dummy.next

Before moving to the next step, let us make sure we also have defined the function `print_linked_list_by_head()` that we introduced earlier in this chapter. Run the following code which is the definition of the said function.

In [5]:
def print_linked_list_by_head(head):
    printout = ''
    node = head
    while node is not None:
        printout += f'{node.val} -> '
        node = node.next
    print(printout)

3.	We want to print both `a_list` and its identical linked list using its head `head_linked_list`. First, run `print(a_list)` and in another chunk run `print_linked_list_by_head(head_linked_list)`. Study the printouts to confirm that `a_list` and its identical linked list do indeed have the same random numbers. 

In [6]:
print(a_list)

[42, 35, 69, 39, 87, 28, 75, 85, 42, 21, 14, 60, 17, 3, 17, 59, 77, 89, 26, 91, 26, 38, 77, 6, 2, 4, 88, 84, 58, 65, 58, 37, 7, 7, 22, 95, 53, 9, 95, 31, 49, 24, 65, 6, 10, 47, 3, 11, 69, 20, 11, 24, 3, 60, 25, 76, 58, 25, 13, 28, 19, 25, 16, 34, 61, 11, 20, 55, 99, 69, 8, 93, 10, 0, 89, 1, 76, 96, 61, 42, 76, 18, 9, 35, 49, 25, 46, 63, 71, 4, 16, 74, 54, 21, 1, 17, 33, 20, 88, 75, 34, 13, 10, 53, 71, 17, 16, 43, 71, 15, 99, 14, 88, 64, 69, 1, 95, 67, 55, 90, 63, 54, 9, 13, 60, 40, 95, 71, 4, 41, 66, 11, 7, 67, 58, 12, 29, 41, 57, 47, 31, 99, 69, 44, 32, 30, 37, 60, 52, 42, 89, 85, 15, 23, 28, 56, 86, 27, 84, 92, 45, 8, 3, 15, 17, 68, 40, 79, 9, 49, 14, 75, 26, 22, 9, 7, 30, 11, 47, 24, 81, 69, 46, 46, 84, 28, 18, 99, 81, 2, 75, 14, 55, 50, 44, 94, 66, 31, 12, 59, 65, 78, 36, 87, 64, 7, 75, 9, 12, 19, 82, 62, 37, 59, 45, 62, 90, 68, 12, 11, 42, 3, 81, 34, 24, 97, 45, 7, 33, 9, 56, 35, 84, 77, 27, 24, 19, 2, 18, 37, 46, 82, 38, 12, 37, 89, 14, 21, 16, 24, 16, 59, 45, 66, 53, 13, 13, 48,

In [7]:
print_linked_list_by_head(head_linked_list)

42 -> 35 -> 69 -> 39 -> 87 -> 28 -> 75 -> 85 -> 42 -> 21 -> 14 -> 60 -> 17 -> 3 -> 17 -> 59 -> 77 -> 89 -> 26 -> 91 -> 26 -> 38 -> 77 -> 6 -> 2 -> 4 -> 88 -> 84 -> 58 -> 65 -> 58 -> 37 -> 7 -> 7 -> 22 -> 95 -> 53 -> 9 -> 95 -> 31 -> 49 -> 24 -> 65 -> 6 -> 10 -> 47 -> 3 -> 11 -> 69 -> 20 -> 11 -> 24 -> 3 -> 60 -> 25 -> 76 -> 58 -> 25 -> 13 -> 28 -> 19 -> 25 -> 16 -> 34 -> 61 -> 11 -> 20 -> 55 -> 99 -> 69 -> 8 -> 93 -> 10 -> 0 -> 89 -> 1 -> 76 -> 96 -> 61 -> 42 -> 76 -> 18 -> 9 -> 35 -> 49 -> 25 -> 46 -> 63 -> 71 -> 4 -> 16 -> 74 -> 54 -> 21 -> 1 -> 17 -> 33 -> 20 -> 88 -> 75 -> 34 -> 13 -> 10 -> 53 -> 71 -> 17 -> 16 -> 43 -> 71 -> 15 -> 99 -> 14 -> 88 -> 64 -> 69 -> 1 -> 95 -> 67 -> 55 -> 90 -> 63 -> 54 -> 9 -> 13 -> 60 -> 40 -> 95 -> 71 -> 4 -> 41 -> 66 -> 11 -> 7 -> 67 -> 58 -> 12 -> 29 -> 41 -> 57 -> 47 -> 31 -> 99 -> 69 -> 44 -> 32 -> 30 -> 37 -> 60 -> 52 -> 42 -> 89 -> 85 -> 15 -> 23 -> 28 -> 56 -> 86 -> 27 -> 84 -> 92 -> 45 -> 8 -> 3 -> 15 -> 17 -> 68 -> 40 -> 79 -> 9 -> 49 -> 14 

**Answer**: Yes, they do.

4.	The following code performs in-place drops of all the even numbers in `a_list`. The code uses the module `time` to record how long it takes for the code to complete. Run the code and note its run time.

In [8]:
import time
t0 = time.time()
i=0
while i<len(a_list):
    if a_list[i] % 2 ==0:
        a_list.pop(i)
    else:
        i+=1        
print(time.time()-t0,'s')

7.483157396316528 s


**Answer**: It took about 7.5 seconds to complete.

5. The following code does the same as we did for `a_list` for its identical linked list. The code is significantly different because the data structures are significantly different. Similar to the previous step, the following code uses the module `time` to record how long it takes for the code to conclude. First, study the code and explain how it drops all the even numbers from the linked list, then run it and note its run time.

``
t0 = time.time()
dummy = ListNode(1)
dummy.next = head_linked_list
node = dummy
previous = node
node = node.next
while node is not None:  
    if node.val %2 ==0:
        previous.next = node.next
    else:
        previous = node
    node = node.next
head_linked_list = dummy.next
print(time.time()-t0,'s')
``

**Answer**: The code gets the CPU to traverse from the head of the linked list to its end. When reaching each node it checks if its val is odd or even. if it is odd, it will continue traversing, but if it is even, it drops the node from the linked list simply by updating the previous.next.

In [10]:
t0 = time.time()
dummy = ListNode(1)
dummy.next = head_linked_list
node = dummy
previous = node
node = node.next
while node is not None:  
    if node.val %2 ==0:
        previous.next = node.next
    else:
        previous = node
    node = node.next
head_linked_list = dummy.next
print(time.time()-t0,'s')

0.019675016403198242 s


**Answer**: it took 20 miliseconds to complete. 

6.	Redo step 3 to make sure that even numbers are indeed dropped from `a_list` and its identical linked_list. 

In [11]:
print(a_list)

[35, 69, 39, 87, 75, 85, 21, 17, 3, 17, 59, 77, 89, 91, 77, 65, 37, 7, 7, 95, 53, 9, 95, 31, 49, 65, 47, 3, 11, 69, 11, 3, 25, 25, 13, 19, 25, 61, 11, 55, 99, 69, 93, 89, 1, 61, 9, 35, 49, 25, 63, 71, 21, 1, 17, 33, 75, 13, 53, 71, 17, 43, 71, 15, 99, 69, 1, 95, 67, 55, 63, 9, 13, 95, 71, 41, 11, 7, 67, 29, 41, 57, 47, 31, 99, 69, 37, 89, 85, 15, 23, 27, 45, 3, 15, 17, 79, 9, 49, 75, 9, 7, 11, 47, 81, 69, 99, 81, 75, 55, 31, 59, 65, 87, 7, 75, 9, 19, 37, 59, 45, 11, 3, 81, 97, 45, 7, 33, 9, 35, 77, 27, 19, 37, 37, 89, 21, 59, 45, 53, 13, 13, 73, 3, 81, 73, 71, 29, 13, 53, 29, 23, 3, 41, 33, 77, 45, 97, 25, 81, 5, 95, 45, 15, 45, 65, 65, 35, 15, 89, 97, 15, 41, 91, 17, 51, 5, 51, 49, 13, 53, 33, 37, 47, 61, 5, 29, 71, 53, 15, 69, 37, 89, 65, 81, 81, 33, 37, 19, 99, 15, 63, 51, 63, 61, 97, 81, 83, 77, 13, 57, 21, 87, 5, 39, 75, 17, 65, 23, 71, 11, 53, 93, 49, 15, 47, 45, 49, 33, 49, 63, 21, 3, 9, 97, 39, 73, 29, 25, 49, 47, 71, 87, 65, 63, 67, 39, 85, 73, 27, 55, 25, 35, 97, 81, 7, 57, 3

In [12]:
print_linked_list_by_head(head_linked_list)

35 -> 69 -> 39 -> 87 -> 75 -> 85 -> 21 -> 17 -> 3 -> 17 -> 59 -> 77 -> 89 -> 91 -> 77 -> 65 -> 37 -> 7 -> 7 -> 95 -> 53 -> 9 -> 95 -> 31 -> 49 -> 65 -> 47 -> 3 -> 11 -> 69 -> 11 -> 3 -> 25 -> 25 -> 13 -> 19 -> 25 -> 61 -> 11 -> 55 -> 99 -> 69 -> 93 -> 89 -> 1 -> 61 -> 9 -> 35 -> 49 -> 25 -> 63 -> 71 -> 21 -> 1 -> 17 -> 33 -> 75 -> 13 -> 53 -> 71 -> 17 -> 43 -> 71 -> 15 -> 99 -> 69 -> 1 -> 95 -> 67 -> 55 -> 63 -> 9 -> 13 -> 95 -> 71 -> 41 -> 11 -> 7 -> 67 -> 29 -> 41 -> 57 -> 47 -> 31 -> 99 -> 69 -> 37 -> 89 -> 85 -> 15 -> 23 -> 27 -> 45 -> 3 -> 15 -> 17 -> 79 -> 9 -> 49 -> 75 -> 9 -> 7 -> 11 -> 47 -> 81 -> 69 -> 99 -> 81 -> 75 -> 55 -> 31 -> 59 -> 65 -> 87 -> 7 -> 75 -> 9 -> 19 -> 37 -> 59 -> 45 -> 11 -> 3 -> 81 -> 97 -> 45 -> 7 -> 33 -> 9 -> 35 -> 77 -> 27 -> 19 -> 37 -> 37 -> 89 -> 21 -> 59 -> 45 -> 53 -> 13 -> 13 -> 73 -> 3 -> 81 -> 73 -> 71 -> 29 -> 13 -> 53 -> 29 -> 23 -> 3 -> 41 -> 33 -> 77 -> 45 -> 97 -> 25 -> 81 -> 5 -> 95 -> 45 -> 15 -> 45 -> 65 -> 65 -> 35 -> 15 -> 89 -> 97 -

**Answer**: Correct! The same odd numbers remain in both data sources.

7.	Compare the runtimes that you recorded for steps 4 and 5. Is there a significant difference? If yes, how can we explain it?

In [13]:
7.483157396316528 / 0.019675016403198242

380.3380512099657

**Answer**: Dropping from `a_list` took about 380 times longer than from its identical linked list.

Yes. The reason is that dropping from a regular list gets the CPU to update the data structure. The CPU has to update the list indexing. However, for the linked list all that CPU needs to do is make a new next assignment.

8.	Summarize your experience from this challenge into one sentence that starts with **“Linked list can be advantageous to big data situations that …”**

**Answer**: *Linked list can be advantageous to big data situations that* we need to make in-place changes regularly to a collection of items that will fall under big data.